### PR TITLE
BACKLOG-12678 : use new routes for content editor

### PR DESCRIPTION
**Pull request by dgriffon**
https://jira.jahia.org/browse/BACKLOG-12678

### CODE REVIEW

File: src/javascript/ContentEditor.jsx [hunk 1]
```diff
@@ -15,9 +13,9 @@ const Modes = {
     create: Create
 };
 
-const ContentEditorCmp = ({mode, path, lang, uilang, site, contentType, client, env, envProps}) => {
+const ContentEditorCmp = ({mode, uuid, lang, uilang, site, contentType, env, envProps}) => {
```
Comment (FBerthelot) [inline src/javascript/ContentEditor.jsx]:
```suggestion
export const ContentEditor = ({mode, uuid, lang, uilang, site, contentType, env, envProps}) => {
```

File: src/javascript/ContentEditor.jsx [hunk 2]
```diff
@@ -30,29 +28,26 @@ const ContentEditorCmp = ({mode, path, lang, uilang, site, contentType, client,
...
-    path: PropTypes.string.isRequired,
+    uuid: PropTypes.string.isRequired,
     lang: PropTypes.string.isRequired,
     uilang: PropTypes.string.isRequired,
     site: PropTypes.string.isRequired,
     contentType: PropTypes.string
 };
 
-export const ContentEditor = withApollo(ContentEditorCmp);
+export const ContentEditor = ContentEditorCmp;
```
Comment (FBerthelot) [inline src/javascript/ContentEditor.jsx]:
```suggestion
```

File: src/javascript/ContentEditor.jsx [hunk 3]
```diff
@@ -30,29 +28,26 @@ const ContentEditorCmp = ({mode, path, lang, uilang, site, contentType, client,
...
 ContentEditor.displayName = 'ContentEditor';
```
Comment (FBerthelot) [inline src/javascript/ContentEditor.jsx]:
```suggestion
```

File: src/javascript/ContentEditorHistory/ContentEditorHistory.jsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import {useHistory} from 'react-router-dom';
+import {splitPath} from './ContentEditorHistory.utils';
+import {Constants} from '~/ContentEditor.constants';
+
+export const useContentEditorHistory = () => {
```
Comment (FBerthelot) [inline src/javascript/ContentEditorHistory/ContentEditorHistory.jsx]: need test :)

File: src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx [hunk 5]
```diff
@@ -1,83 +1,83 @@
...
+            componentRenderer.destroy('CreateNewContentDialog');
+        };
 
-const mapStateToProps = state => {
-    return {
-        uilang: state.uilang,
-        language: state.language,
-        siteKey: state.site
-    };
+        componentRenderer.render(
```
Comment (FBerthelot) [inline src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx]: We can go deeper and render the modal directly in the component 🤔 
But as we don't have modal in moonstone yet, I'm to keep it like this.

File: src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx [hunk 6]
```diff
@@ -1,83 +1,83 @@
...
+        context.showOnNodeTypes,
+        transformNodeTypesToActions);
+
+    if (Loading && (loadingTypes || res.loading)) {
+        return <Loading context={context}/>;
+    }
+
+    if (error) {
+        console.error(error);
+        return <>{error}</>;
```
Comment (FBerthelot) [inline src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx]: We have a better error message to display to user now. You can use it.

File: src/javascript/Edit/EditContent.action.jsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import * as PropTypes from 'prop-types';
+import {useSelector} from 'react-redux';
+import {Constants} from '~/ContentEditor.constants';
+
+export const EditContent = ({context, render: Render, loading: Loading}) => {
+    const {redirect} = useContentEditorHistory();
+    const language = useSelector(state => state.language);
+    const res = useNodeInfo({path: context.path, language}, {getDisplayName: true});
+
+    if (res.loading && Loading && !res.node) {
```
Comment (FBerthelot) [inline src/javascript/Edit/EditContent.action.jsx]:
```suggestion
    if (res.loading && !res.node) {
```
loading is already required by the component in the prop-types, no need to double check.

It's really necessary to test if there is no res.node ?

File: src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx (hunk 5 above)
Comment (dgriffon) [reply src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx]: thanks for the advice!

File: src/javascript/ContentEditor.jsx (hunk 2 above)
Comment (dgriffon) [reply src/javascript/ContentEditor.jsx]: done

File: src/javascript/Api/ContentEditor.api.jsx [hunk 8]
```diff
@@ -53,10 +54,10 @@ const ContentEditorApiCmp = ({classes, client}) => {
      * @param excludedNodeTypes (optional) The node types excluded for creation, by default: ['jmix:studioOnly', 'jmix:hiddenType']
      * @param includeSubTypes (optional) if true, subtypes of nodeTypes provided will be resolved.
      */
-    window.CE_API.create = (path, site, lang, uilang, nodeTypes, excludedNodeTypes, includeSubTypes) => {
+    window.CE_API.create = (uuid, path, site, lang, uilang, nodeTypes, excludedNodeTypes, includeSubTypes) => {
```
Comment (FBerthelot) [inline src/javascript/Api/ContentEditor.api.jsx]: There is possibility to do that ? 
```suggestion
    window.CE_API.create = ({uuid, path, site, lang, uilang, nodeTypes, excludedNodeTypes, includeSubTypes}) => {
```

File: src/javascript/ContentEditorHistory/ContentEditorHistory.jsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        let newRest = rest || (rest === '' ? '' : oldRest);
+        // Add starting "/" if necessary
+        newRest = newRest && (newRest.startsWith('/') ? newRest : ('/' + newRest));
+        history.push(buildContentEditorURL(Constants.appName, mode || currentMode, language || currentLanguage, uuid || currentUuid, newRest));
+    };
+
+    return {redirect};
+};
+
+const buildContentEditorURL = (appName, mode, language, uuid, rest) => {
```
Comment (FBerthelot) [inline src/javascript/ContentEditorHistory/ContentEditorHistory.jsx]:
```suggestion
const buildContentEditorURL = ({appName, mode, language, uuid, rest}) => {
```
Same question :)

File: src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx [hunk 10]
```diff
@@ -1,83 +1,92 @@
...
+import {ComponentRendererContext} from '@jahia/ui-extender';
+import * as PropTypes from 'prop-types';
+import {useContentEditorHistory} from '~/ContentEditorHistory';
+import {useTranslation} from 'react-i18next';
 
-// Should continue to use redux here, because this action is used to replace JContent create action, so it's only displayed in JContent
-const mapDispatchToProps = dispatch => ({
-    gotToCE: gotoParams => dispatch(cmGoto(gotoParams))
-});
+const onClick = (uuid, language, context, redirect, componentRenderer) => {
```
Comment (FBerthelot) [inline src/javascript/Create/CreateNewContentAction/createNewContent.action.jsx]:
```suggestion
const onClick = ({uuid, language, context, redirect, componentRenderer}) => {
```
Same question

File: src/javascript/Create/CreateNewContentAction/createNewContent.utits.js [hunk 11]
```diff
@@ -1,13 +1,43 @@
 import {getTreeOfContentWithRequirements} from './CreateNewContent.gql-queries';
+import {useQuery} from '@apollo/react-hooks';
 
 const NB_OF_DISPLAYED_RESCRICTED_SUB_NODES = 5;
 
+export const useCreatableNodetypes = (nodeTypes, includeSubTypes, path, uilang, excludedNodeTypes, showOnNodeTypes, transformResultCallback) => {
```
Comment (FBerthelot) [inline src/javascript/Create/CreateNewContentAction/createNewContent.utits.js]:
```suggestion
export const useCreatableNodetypes = ({nodeTypes, includeSubTypes, path, uilang, excludedNodeTypes, showOnNodeTypes, transformResultCallback}) => {
```
Same question

File: src/javascript/Api/ContentEditor.api.jsx (hunk 8 above)
Comment (dgriffon) [reply src/javascript/Api/ContentEditor.api.jsx]: It would have been nice, but as an exposed API, it is better to know the parameters to pass

**Pull request merged**